code/modules/paperwork/clipboard.dm

```diff
@@ -8,119 +11,167 @@
-
-		for(P in src)
-			if(P == toppaper)
-				continue
-			dat += "<A href='?src=[REF(src)];write=[REF(P)]'>Write</A> <A href='?src=[REF(src)];remove=[REF(P)]'>Remove</A> <A href='?src=[REF(src)];top=[REF(P)]'>Move to top</A> - <A href='?src=[REF(src)];read=[REF(P)]'>[P.name]</A><BR>"
-	user << browse(dat, "window=clipboard")
-	onclose(user, "clipboard")
 	add_fingerprint(usr)
+	ui_interact(user)
+	return
+
+/obj/item/clipboard/ui_interact(mob/user, datum/tgui/ui)
+	ui = SStgui.try_update_ui(user, src, ui)
+	if(!ui)
+		ui = new(user, src, "Clipboard")
+		ui.open()
+
+/obj/item/clipboard/ui_data(mob/user)
+	// prepare data for TGUI
+	var/list/data = list()
+	data["pen"] = "[pen]"
+
+	var/obj/item/paper/P = toppaper
+	data["top_paper"] = "[P]"
+	data["top_paper_ref"] = "[REF(P)]"
+
+	data["paper"] = list()
+	data["paper_ref"] = list()
+	for(P in src)
+		if(P == toppaper)
+			continue
+		data["paper"] += "[P]"
+		data["paper_ref"] += "[REF(P)]"
+
+	return data
+
+/obj/item/clipboard/ui_act(action, params)
+	. = ..()
+	if(.)
+		return
 
-
-/obj/item/clipboard/Topic(href, href_list)
-	..()
 	if(usr.stat != CONSCIOUS || HAS_TRAIT(usr, TRAIT_HANDS_BLOCKED))
 		return
 
-	if(usr.contents.Find(src))
-
-		if(href_list["pen"])
-			if(haspen)
-				haspen.forceMove(usr.loc)
-				usr.put_in_hands(haspen)
-				haspen = null
-
-		if(href_list["addpen"])
-			if(!haspen)
-				var/obj/item/held = usr.get_active_held_item()
-				if(istype(held, /obj/item/pen))
-					var/obj/item/pen/W = held
-					if(!usr.transferItemToLoc(W, src))
-						return
-					haspen = W
-					to_chat(usr, "<span class='notice'>You slot [W] into [src].</span>")
-
-		if(href_list["write"])
-			var/obj/item/P = locate(href_list["write"]) in src
+	switch(action)
+		// Take the pen out
+		if("remove_pen")
+			if(pen)
+				remove_pen(usr)
+				. = TRUE
+		// Take paper out
+		if("remove_paper")
+			var/obj/item/paper/P = locate(params["ref"]) in src
 			if(istype(P))
-				if(usr.get_active_held_item())
-					P.attackby(usr.get_active_held_item(), usr)
-
-		if(href_list["remove"])
-			var/obj/item/P = locate(href_list["remove"]) in src
-			if(istype(P))
-				P.forceMove(usr.loc)
-				usr.put_in_hands(P)
-				if(P == toppaper)
-					toppaper = null
-					var/obj/item/paper/newtop = locate(/obj/item/paper) in src
-					if(newtop && (newtop != P))
-						toppaper = newtop
-					else
-						toppaper = null
-
-		if(href_list["read"])
-			var/obj/item/paper/P = locate(href_list["read"]) in src
+				remove_paper(P, usr)
+				. = TRUE
+		// Look at (or edit) the paper
+		if("edit_paper")
+			var/obj/item/paper/P = locate(params["ref"]) in src
 			if(istype(P))
-				usr.examinate(P)
-
-		if(href_list["top"])
-			var/obj/item/P = locate(href_list["top"]) in src
+				P.ui_interact(usr)
+				update_icon()
+				. = TRUE
+		// Move paper to the top
+		if("move_top_paper")
+			var/obj/item/paper/P = locate(params["ref"]) in src
```

Comment:
https://github.com/tgstation/tgstation/blob/master/.github/CONTRIBUTING.md#use-descriptive-and-obvious-names